File: sdks/go/protocol.go
package emergent

import (
	"encoding/binary"
	"encoding/json"
	"fmt"

	"github.com/vmihailenco/msgpack/v5"
)

// Protocol constants matching acton-reactive IPC.
const (
	ProtocolVersion byte = 0x02
	HeaderSize      int  = 7
	MaxFrameSize    int  = 16 * 1024 * 1024 // 16 MiB
)

// IPC message type constants.
const (
	MsgTypeRequest     byte = 0x01
	MsgTypeResponse    byte = 0x02
	MsgTypeError       byte = 0x03
	MsgTypeHeartbeat   byte = 0x04
	MsgTypePush        byte = 0x05
	MsgTypeSubscribe   byte = 0x06
	MsgTypeUnsubscribe byte = 0x07
	MsgTypeDiscover    byte = 0x08
	MsgTypeStream      byte = 0x09

	// MsgTypeSubscribePatterns subscribes to IPC prefix patterns.
	// Requires an engine newer than 0.10.10.
	MsgTypeSubscribePatterns byte = 0x0a
	// MsgTypeUnsubscribePatterns unsubscribes from IPC prefix patterns.
	// Requires an engine newer than 0.10.10.
	MsgTypeUnsubscribePatterns byte = 0x0b
)

// Serialization format constants.
const (
	FormatJSON    byte = 0x01
	FormatMsgPack byte = 0x02
)

// DecodedFrame represents a decoded protocol frame.
type DecodedFrame struct {
	MsgType       byte
	Format        byte
	Payload       any
	BytesConsumed int
}

// EncodeFrame encodes a payload into a protocol frame.
//
// Frame structure:
//   - [0-3]: Payload length (big-endian u32)
//   - [4]: Protocol version
//   - [5]: Message type
//   - [6]: Serialization format
//   - [7+]: Payload bytes
func EncodeFrame(msgType byte, payload any, format byte) ([]byte, error) {
	var payloadBytes []byte
	var err error

	switch format {
	case FormatMsgPack:
		payloadBytes, err = msgpack.Marshal(payload)
	case FormatJSON:
		payloadBytes, err = json.Marshal(payload)
	default:
		return nil, &ProtocolError{Msg: fmt.Sprintf("unsupported format: %d", format)}
	}
	if err != nil {
		return nil, fmt.Errorf("serialization error: %w", err)
	}

	payloadLen := len(payloadBytes)
	if payloadLen > MaxFrameSize {
		return nil, &ProtocolError{
			Msg: fmt.Sprintf("payload too large: %d bytes (max: %d)", payloadLen, MaxFrameSize),
		}
	}

	frame := make([]byte, HeaderSize+payloadLen)
	binary.BigEndian.PutUint32(frame[0:4], uint32(payloadLen))
	frame[4] = ProtocolVersion
	frame[5] = msgType
	frame[6] = format
	copy(frame[HeaderSize:], payloadBytes)

	return frame, nil
}

// frameStepKind says what the front of a read buffer holds.
type frameStepKind int

const (
	// frameIncomplete: not a whole frame yet, so wait for more bytes.
	frameIncomplete frameStepKind = iota
	// frameDecoded: a decoded frame.
	frameDecoded
	// frameBadBody: a whole frame whose body does not decode. Its length is
	// known, so the reader drops bytesConsumed bytes and carries on.
	frameBadBody
	// frameBadFraming: a header that cannot be trusted, so the reader cannot
	// tell where the next frame starts.
	frameBadFraming
)

// frameStep is the result of reading the front of a read buffer.
type frameStep struct {
	kind frameStepKind
	// frame is set for frameDecoded.
	frame *DecodedFrame
	// msgType and bytesConsumed are set for frameBadBody.
	msgType       byte
	bytesConsumed int
	// reason is set for frameBadBody and frameBadFraming.
	reason string
}

// nextFrame reads the frame at the front of a buffer. A body that does not
// decode is told apart from a header that cannot be trusted, because only the
// first leaves the reader able to find the next frame.
func nextFrame(buffer []byte) frameStep {
	if len(buffer) < HeaderSize {
		return frameStep{kind: frameIncomplete} // Not enough data for header
	}

	payloadLen := binary.BigEndian.Uint32(buffer[0:4])
	if int(payloadLen) > MaxFrameSize {
		return frameStep{kind: frameBadFraming, reason: fmt.Sprintf("frame too large: %d bytes", payloadLen)}
	}

	totalLen := HeaderSize + int(payloadLen)
	if len(buffer) < totalLen {
		return frameStep{kind: frameIncomplete} // Not enough data for full frame
	}

	version := buffer[4]
	if version != ProtocolVersion {
		return frameStep{
			kind:   frameBadFraming,
			reason: fmt.Sprintf("unsupported protocol version: %d (expected %d)", version, ProtocolVersion),
		}
	}

	msgType := buffer[5]
	format := buffer[6]
	payloadBytes := buffer[HeaderSize:totalLen]

	// A heartbeat is a bare header. Its body is empty, which neither format
	// can decode, so it is returned with a nil payload.
	if payloadLen == 0 {
		return frameStep{
			kind:  frameDecoded,
			frame: &DecodedFrame{MsgType: msgType, Format: format, BytesConsumed: totalLen},
		}
	}

	badBody := func(reason string) frameStep {
		return frameStep{kind: frameBadBody, msgType: msgType, bytesConsumed: totalLen, reason: reason}
	}

	var payload any
	var err error

	switch format {
	case FormatMsgPack:
		err = msgpack.Unmarshal(payloadBytes, &payload)
	case FormatJSON:
		err = json.Unmarshal(payloadBytes, &payload)
	default:
		return badBody(fmt.Sprintf("unknown format: %d", format))
	}
	if err != nil {
		return badBody(fmt.Sprintf("deserialization error: %v", err))
	}

	return frameStep{
		kind: frameDecoded,
		frame: &DecodedFrame{
			MsgType:       msgType,
			Format:        format,
			Payload:       payload,
			BytesConsumed: totalLen,
		},
	}
}

// TryDecodeFrame tries to decode a frame from a buffer.
// Returns nil if the buffer does not contain a complete frame.
func TryDecodeFrame(buffer []byte) (*DecodedFrame, error) {
	step := nextFrame(buffer)
	switch step.kind {
	case frameDecoded:
		return step.frame, nil
	case frameBadBody, frameBadFraming:
		return nil, &ProtocolError{Msg: step.reason}
	default:
		return nil, nil
	}
}

File: sdks/ts/src/protocol.ts
/**
 * IPC protocol implementation matching acton-reactive.
 * @module
 */

import { decode, encode } from "@msgpack/msgpack";
import { ProtocolError } from "./errors.ts";

// ============================================================================
// Protocol Constants (matching acton-reactive IPC)
// ============================================================================

/** Protocol version */
export const PROTOCOL_VERSION = 0x02;

/** Maximum frame size (16 MiB) */
export const MAX_FRAME_SIZE = 16 * 1024 * 1024;

/** Frame header size: length(4) + version(1) + msgType(1) + format(1) */
export const HEADER_SIZE = 7;

// Message types (matching acton-reactive/src/common/ipc/protocol.rs)
export const MSG_TYPE_REQUEST = 0x01;
export const MSG_TYPE_RESPONSE = 0x02;
export const MSG_TYPE_ERROR = 0x03;
export const MSG_TYPE_HEARTBEAT = 0x04;
export const MSG_TYPE_PUSH = 0x05;
export const MSG_TYPE_SUBSCRIBE = 0x06;
export const MSG_TYPE_UNSUBSCRIBE = 0x07;
export const MSG_TYPE_DISCOVER = 0x08;
export const MSG_TYPE_STREAM = 0x09;
/** Subscribe to IPC prefix patterns (requires engine 0.14.0 or later). */
export const MSG_TYPE_SUBSCRIBE_PATTERNS = 0x0a;
/** Unsubscribe from IPC prefix patterns (requires engine 0.14.0 or later). */
export const MSG_TYPE_UNSUBSCRIBE_PATTERNS = 0x0b;

// Serialization formats
export const FORMAT_JSON = 0x01;
export const FORMAT_MSGPACK = 0x02;

// ============================================================================
// Encoding/Decoding
// ============================================================================

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

/**
 * Encode a frame for transmission.
 *
 * Frame structure:
 * - [0-3]: Payload length (big-endian u32)
 * - [4]: Protocol version
 * - [5]: Message type
 * - [6]: Serialization format
 * - [7+]: Payload bytes
 */
export function encodeFrame(
  msgType: number,
  payload: unknown,
  format = FORMAT_MSGPACK,
): Uint8Array {
  let payloadBytes: Uint8Array;

  if (format === FORMAT_MSGPACK) {
    payloadBytes = encode(payload);
  } else if (format === FORMAT_JSON) {
    const jsonStr = JSON.stringify(payload);
    payloadBytes = textEncoder.encode(jsonStr);
  } else {
    throw new ProtocolError(`Unsupported format: ${format}`);
  }

  const payloadLen = payloadBytes.length;

  if (payloadLen > MAX_FRAME_SIZE) {
    throw new ProtocolError(
      `Payload too large: ${payloadLen} bytes (max: ${MAX_FRAME_SIZE})`,
    );
  }

  const frame = new Uint8Array(HEADER_SIZE + payloadLen);
  const view = new DataView(frame.buffer);

  // Header
  view.setUint32(0, payloadLen, false); // big-endian
  frame[4] = PROTOCOL_VERSION;
  frame[5] = msgType;
  frame[6] = format;

  // Payload
  frame.set(payloadBytes, HEADER_SIZE);

  return frame;
}

/**
 * Decode a frame body.
 *
 * An empty body decodes to `null`. A heartbeat frame is a bare header, and
 * neither JSON nor MessagePack can parse zero bytes.
 *
 * @throws {ProtocolError} If the format byte is unknown, or the body is not
 *   valid in its format, such as one cut short
 */
export function decodePayload(
  payloadBytes: Uint8Array,
  format: number,
): unknown {
  if (format !== FORMAT_MSGPACK && format !== FORMAT_JSON) {
    throw new ProtocolError(`Unknown format: ${format}`);
  }

  if (payloadBytes.length === 0) {
    return null;
  }

  // Both decoders throw their own errors, a SyntaxError from JSON and a
  // RangeError or DecodeError from MessagePack. Callers handle one class.
  try {
    if (format === FORMAT_JSON) {
      return JSON.parse(textDecoder.decode(payloadBytes));
    }
    return decode(payloadBytes);
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err);
    const name = format === FORMAT_JSON ? "JSON" : "MessagePack";
    throw new ProtocolError(`Malformed ${name} frame body: ${reason}`);
  }
}

/** Names of the frame types this SDK knows, keyed by their wire byte. */
const FRAME_TYPE_NAMES: Readonly<Record<number, string>> = {
  [MSG_TYPE_REQUEST]: "REQUEST",
  [MSG_TYPE_RESPONSE]: "RESPONSE",
  [MSG_TYPE_ERROR]: "ERROR",
  [MSG_TYPE_HEARTBEAT]: "HEARTBEAT",
  [MSG_TYPE_PUSH]: "PUSH",
  [MSG_TYPE_SUBSCRIBE]: "SUBSCRIBE",
  [MSG_TYPE_UNSUBSCRIBE]: "UNSUBSCRIBE",
  [MSG_TYPE_DISCOVER]: "DISCOVER",
  [MSG_TYPE_STREAM]: "STREAM",
  [MSG_TYPE_SUBSCRIBE_PATTERNS]: "SUBSCRIBE_PATTERNS",
  [MSG_TYPE_UNSUBSCRIBE_PATTERNS]: "UNSUBSCRIBE_PATTERNS",
};

/**
 * Name a frame type byte for a log line.
 *
 * Returns `undefined` for a byte this SDK does not know.
 */
export function frameTypeName(msgType: number): string | undefined {
  return FRAME_TYPE_NAMES[msgType];
}

/**
 * Decoded frame result.
 */
export interface DecodedFrame {
  /** Message type constant */
  msgType: number;
  /** Serialization format */
  format: number;
  /** Deserialized payload */
  payload: unknown;
  /** Total bytes consumed from buffer */
  bytesConsumed: number;
}

/**
 * What the front of a read buffer holds.
 *
 * - `incomplete`: not a whole frame yet, so wait for more bytes
 * - `frame`: a decoded frame
 * - `bad-body`: a whole frame whose body does not decode. Its length is known,
 *   so the reader drops `bytesConsumed` bytes and carries on with the next
 *   frame
 * - `bad-framing`: a header that cannot be trusted, so the reader cannot tell
 *   where the next frame starts
 */
export type FrameStep =
  | { readonly kind: "incomplete" }
  | { readonly kind: "frame"; readonly frame: DecodedFrame }
  | {
    readonly kind: "bad-body";
    readonly msgType: number;
    readonly bytesConsumed: number;
    readonly reason: string;
  }
  | { readonly kind: "bad-framing"; readonly reason: string };

/**
 * Read the frame at the front of a buffer without throwing.
 *
 * A body that does not decode is told apart from a header that cannot be
 * trusted, because only the first leaves the reader able to find the next
 * frame.
 */
export function nextFrame(buffer: Uint8Array): FrameStep {
  if (buffer.length < HEADER_SIZE) {
    return { kind: "incomplete" };
  }

  const view = new DataView(
    buffer.buffer,
    buffer.byteOffset,
    buffer.byteLength,
  );
  const payloadLen = view.getUint32(0, false); // big-endian

  if (payloadLen > MAX_FRAME_SIZE) {
    return {
      kind: "bad-framing",
      reason: `Frame too large: ${payloadLen} bytes`,
    };
  }

  const version = buffer[4];
  if (version !== PROTOCOL_VERSION) {
    return {
      kind: "bad-framing",
      reason:
        `Unsupported protocol version: ${version} (expected ${PROTOCOL_VERSION})`,
    };
  }

  const totalLen = HEADER_SIZE + payloadLen;

  if (buffer.length < totalLen) {
    return { kind: "incomplete" };
  }

  const msgType = buffer[5];
  const format = buffer[6];

  try {
    const payload = decodePayload(
      buffer.subarray(HEADER_SIZE, totalLen),
      format,
    );
    return {
      kind: "frame",
      frame: { msgType, format, payload, bytesConsumed: totalLen },
    };
  } catch (err) {
    return {
      kind: "bad-body",
      msgType,
      bytesConsumed: totalLen,
      reason: err instanceof Error ? err.message : String(err),
    };
  }
}

/**
 * Try to decode a frame from a buffer.
 *
 * Returns null if the buffer doesn't contain a complete frame.
 * Throws ProtocolError if the frame is malformed.
 */
export function tryDecodeFrame(buffer: Uint8Array): DecodedFrame | null {
  const step = nextFrame(buffer);
  switch (step.kind) {
    case "incomplete":
      return null;
    case "frame":
      return step.frame;
    case "bad-body":
    case "bad-framing":
      throw new ProtocolError(step.reason);
  }
}

import { typeid } from "typeid-js";

/**
 * Generate a unique correlation ID in TypeID format.
 */
export function generateCorrelationId(prefix = "req"): string {
  return typeid(prefix).toString();
}
